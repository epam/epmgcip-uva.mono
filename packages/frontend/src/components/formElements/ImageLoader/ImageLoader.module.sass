@use 'src/style/colors'
@import 'src/style/typography'

.imageLoaderWrapper
    width: 100%
    margin-bottom: 18px

    .imageLoader
        width: 100%
        height: 180px
        background-color: colors.$gray-200

        .imageLoaderContainer
            display: flex
            flex-direction: row
            justify-content: end
            align-items: flex-end
            height: 100%

            .downloadButton
                position: absolute
                z-index: 1
                background-color: colors.$white
                border-radius: 8px
                border: 1px solid colors.$blue-dark
                padding: 15px 14px
                margin: 8px 20px
                cursor: pointer

            .imageLoaderInput
                display: none

            .imagePreview
                width: 100%
                height: 100%
                object-fit: cover

.validationError
    @extend .T3
    align-self: flex-start
    color: colors.$red
    margin-top: 2px
    margin-bottom: 0
    padding: 0 24px
