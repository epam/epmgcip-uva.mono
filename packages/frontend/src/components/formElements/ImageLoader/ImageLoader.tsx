import { getClassesList } from 'src/utils/getClassesList';
import css from './ImageLoader.module.sass';
import { useState } from 'react';
import CameraSvg from './assets/camera.svg';
import { showNotification } from 'src/utils/showNotification';
import { IMAGE_TYPE, NOTIFICATIONS } from 'src/constants';

interface ImageLoaderProps {
  setImage: React.Dispatch<React.SetStateAction<File | null>>;
  isValidationError?: boolean;
  errorMessage?: string;
  previewExistingImage?: string;
}

const convertBytesToMegabytes = (value: number) =>
  Number((value / (1024 * 1024)).toFixed(4));

export const ImageLoader = ({
  setImage,
  isValidationError,
  errorMessage,
  previewExistingImage
}: ImageLoaderProps) => {
  const imageLoaderClasses = getClassesList(css.imageLoader);
  const [preview, setPreview] = useState<string>(previewExistingImage ? previewExistingImage : '');
  

  const handleSetImage = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files !== null) {
      const imageFile = e.target.files[0];

      if (convertBytesToMegabytes(imageFile.size) > 10) {
        showNotification(NOTIFICATIONS().IMAGE_TOO_BIG, 6000);
        return;
      }

      if (!IMAGE_TYPE.some((type) => type === imageFile.type)) {
        showNotification(NOTIFICATIONS().IMAGE_WRONG_TYPE, 6000);
        return;
      }

      setImage(() => imageFile);
      setPreview(() => URL.createObjectURL(imageFile));
    }
  };

  return (
    <div className={css.imageLoaderWrapper}>
      <div className={imageLoaderClasses}>
        <label className={css.imageLoaderContainer}>
          <img className={css.downloadButton} width={19} src={CameraSvg} />
          <input
            className={css.imageLoaderInput}
            type='file'
            onChange={(e) => handleSetImage(e)}
          />
          {preview && (
            <img
              className={css.imagePreview}
              src={preview}
              alt='Your Image Preview'
            />
          )}
        </label>
      </div>
      {isValidationError && (
        <p className={css.validationError}>
          {isValidationError && errorMessage}
        </p>
      )}
    </div>
  );
};
