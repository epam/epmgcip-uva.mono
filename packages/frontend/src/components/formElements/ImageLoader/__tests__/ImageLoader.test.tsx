import { describe, expect, it, vi } from 'vitest';
import { render } from '@testing-library/react';
import { ImageLoader } from '../ImageLoader';

const setImageMock = () => vi.fn();

describe('Testing: ImageLoader', () => {
  it('should render', () => {
    const { container } = render(<ImageLoader setImage={setImageMock} />);

    expect(container).toMatchSnapshot();
  });
});
