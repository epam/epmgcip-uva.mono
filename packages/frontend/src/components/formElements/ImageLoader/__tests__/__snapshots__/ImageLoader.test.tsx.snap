// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing: ImageLoader > should render 1`] = `
<div>
  <div
    class="_imageLoaderWrapper_112c4b"
  >
    <div
      class="_imageLoader_112c4b"
    >
      <label
        class="_imageLoaderContainer_112c4b"
      >
        <img
          class="_downloadButton_112c4b"
          src="/packages/frontend/src/components/formElements/ImageLoader/assets/camera.svg"
          width="19"
        />
        <input
          class="_imageLoaderInput_112c4b"
          type="file"
        />
      </label>
    </div>
  </div>
</div>
`;
